### PR TITLE
Add aws_ses_identity_policy module for managing SES sending policies

### PR DESCRIPTION
**Pull request by orthanc**
##### SUMMARY
Introduce aws_ses_identity_policy_module. This module allows the user to manage sending authorization policies associated with an SES identity (email or domain).

SES authorization sending policies can be used to control what actors are able to send email on behalf of the validated identity and what conditions must be met by the sent emails.

Examples:

    # Note: These examples do not set authentication details, see the AWS Guide for details.

    - name: add sending authorization policy to domain identity
    aws_ses_identity_policy:
        identity: example.com
        policy_name: ExamplePolicy
        policy: "{{ lookup('template', 'policy.json.j2') }}"
        state: present

    - name: add sending authorization policy to email identity
    aws_ses_identity_policy:
        identity: example@example.com
        policy_name: ExamplePolicy
        policy: "{{ lookup('template', 'policy.json.j2') }}"
        state: present

    - name: add sending authorization policy to identity using ARN
    aws_ses_identity_policy:
        identity: "arn:aws:ses:us-east-1:12345678:identity/example.com"
        policy_name: ExamplePolicy
        policy: "{{ lookup('template', 'policy.json.j2') }}"
        state: present

    - name: remove sending authorization policy
    aws_ses_identity_policy:
        identity: example.com
        policy_name: ExamplePolicy
        state: absent

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aws_ses_identity_policy

##### ANSIBLE VERSION
```
ansible 2.6.0 (aws_ses_identity_policy_module 1caa87705e) last updated 2018/02/23 22:05:12 (GMT +1300)
  config file = None
  configured module search path = [u'USER_HOME_DIR/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = USER_HOME_DIR/projects/ansible/lib/ansible
  executable location = ANSIBLE_CHECKOUT_LOCATION/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```

##### ADDITIONAL INFORMATION

This builds on top of the `aws_ses_identity` module added in 2.5 to add the ability to control the identity policies.

The integration tests require additional AWS permissions which have been added in `hacking/aws_config/testing_policies/compute-policy.json`.

Because of this I suspect the shippable build will fail till this permission is granted for the builds. Assuming this is the case I'll add a second commit to disable the test until the permissions are added.

### CODE REVIEW

**Comment by orthanc** [comment]
It may be appropriate to split this PR into two seperate changes, but I've left together for now since if split one half would seem almost a "style only" PR.

The main chage is described in the PR description. But at present there's an error in the common aws doc fragment that fails one of the [new checks](https://github.com/ansible/ansible/commit/50adc5409bd23051dec718a0442f9075b13b4696). So can either continue the pattern of ignoring this failure in the new module or just fix it in the common include and un-ignore the 205 other affected modules.

The second commit on this PR is fixing this common error and un-ignoring the now fixed modules. It's essentially unrelated tidy up, but something is needed in order for the sanity tests to pass with this new module.

If it's preferrable to just ignore the error here or split this off into a seperate PR let me know and I'll action it.

**Comment by orthanc** [comment]
Also note that as mentioned in description, the integration test currently requires additional permissions. So it's been disabled in order to get a passing build. Please revert or remove the final commit to re-enable this once appropriate permissions have been granted to the shippable build.

**Comment by mattclay** [comment]
@orthanc Please split the docs fragments fix into a separate PR and CC me. Also, re-enable the integration tests and I'll mark the PR as needing a CI permissions update. Then we can re-run CI on the PR when the required permissions are available.

**Comment by orthanc** [comment]
@mattclay Done.

[pull/36653](https://github.com/ansible/ansible/pull/36653) is the PR for the doc fragment update. I've stripped that out of this PR.

Integration test is also re-enabled now in this PR

**Comment by orthanc** [comment]
ready_for_review

Latest commits update to use AnsibleAWSModule, I've had to extend the connection helpers to allow passing extra parameters through so I can set the botocore config due to the low throttling thresholds on SES.

Remaining test failure is just due to missing permissions in the CI build.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+def create_or_update_identity_policy(connection, module):
+    identity = module.params.get('identity')
+    policy_name = module.params.get('policy_name')
+    required_policy = module.params.get('policy')
+    required_policy_dict = sort_json_policy_dict(json.loads(required_policy))
+
+    changed = False
+    policy = get_identity_policy(connection, module, identity, policy_name)
+    policy_dict = sort_json_policy_dict(json.loads(policy)) if policy else None
+    if required_policy_dict != policy_dict:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py]: [compare_policies()](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py#L681) is a better way of checking if there are changes (it takes two dictionaries, recursively sorts and makes them hashable for comparison and returns True if they are different). s3_bucket and iam_managed_policy were both using sort_json_policy_dict which introduced bugs.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    )
+
+    # Allow up to 10 attempts to call the SES APIs before giving up (9 retries).
+    # SES APIs seem to have a much lower throttling threshold than most of the rest of the AWS APIs.
+    # Docs say 1 call per second. This shouldn't actually be a big problem for normal usage, but
+    # the ansible build runs multiple instances of the test in parallel.
+    # As a result there are build failures due to throttling that exceeds boto's default retries.
+    # The back-off is exponential, so upping the retry attempts allows multiple parallel runs
+    # to succeed.
+    boto_core_config = Config(retries={'max_attempts': 9})
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py]: Is there a reason not to use AWSRetry? It also has jittered_backoff which is nice.

**Review by s-hertel — COMMENTED** [review]
Updated CI. This looks great. Thanks, @orthanc.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py (hunk 2 above)
Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py]: There is a small benefit of this approach. Since the low throttling limits seem to apply to just about all the SES APIs, doing it with botocore config means it applies across the board. Where as AWSRetry has to be applied at each callsite.

To use as a decorator I'll have to pull out every boto call into a method while still doing the error handling in the calling method. I think this might get sufficiently verbose that it would obscure the function. 

But the jittered backoff does look nice, I'll have another look at how to use AWSRetry at all calls without making it hard to follow the code. 

Would you be open to an extension of AWSRetry that wraps a boto client and automatically decorates all the calls?

I'm not not sure if I can make that work. But if I can it feels like a nice way of handling this case.  Unless you feel that would add confusing complexity. 

If I can't make a general wrapping work, or you'd prefer to avoid that approach,  I'll go through and decorate all the call sites with AWSRetry.

**Comment by orthanc** [comment]
ready_for_review

I've switched over to compare_policies as suggested.

I've replaced the use of boto-core config, I've done this by adding an extra parameter to AnsibleAWSModule connection methods to return a wrapped connection that applies the retry decorator to every method.

The rationale for this is that because of the low limits on the SES APIs I think we need the retry on every call to avoid the shippbable failres from the parallel builds.

As I said in the previous comment, if the maintainers feel this is just over-complicating it I'll switch it over to just simple decorators.

**Comment by orthanc** [comment]
ready_for_review

**Comment by orthanc** [comment]
bot_status

File: lib/ansible/module_utils/aws/core.py [hunk 3]
```diff
@@ -158,3 +160,18 @@ def fail_json_aws(self, exception, msg=None):
         else:
             self._module.fail_json(msg=message, exception=last_traceback,
                                    **camel_dict_to_snake_dict(response))
+
+
+class _RetryingBotoClientWrapper:
```
Comment (s-hertel) [inline lib/ansible/module_utils/aws/core.py]: This is kind of cool. I like it. @ryansb @willthames Do you have thoughts on this?

It should be an new-style class `class _RetryingBotoClientWrapper(object):`

The one possible pitfall of this (that I can think of) is that if people see it as a shortcut but need to not retry on a specific exception it would be easy to overlook and retry inefficiently but that would probably be an uncommon problem and just something to keep an eye out for when reviewing things that use this.

File: lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    if compare_policies(policy_dict, required_policy_dict):
+        changed = True
+        try:
+            if not module.check_mode:
+                connection.put_identity_policy(Identity=identity, PolicyName=policy_name, Policy=required_policy)
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg='Failed to put identity policy {policy}'.format(policy=policy_name))
+
+    # Load the list of applied policies to include in the response.
+    # In principle we should be able to just return the response, but given
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_ses_identity_policy.py:122]: Yeah, it looks like the response for put_identity_policy is only the response metadata, so listing the identity policies looks like the right approach.

**Review by s-hertel — APPROVED** [review]

File: lib/ansible/module_utils/aws/core.py (hunk 3 above)
Comment (willthames) [reply lib/ansible/module_utils/aws/core.py]: It's pretty magical (in both good and bad senses!). 

Your worry is reasonable, I've accidentally hit infinite loops even when knowing I'm hitting AWSRetry, this might definitely cause some problems for module development down the line. I do like the idea of not having to write a separate method just for backoffs though!

Comment (orthanc) [reply lib/ansible/module_utils/aws/core.py]: What about if we make it more explicit at the call-site without still requiring a wrapping method. E.g. something like:

    # Setup Connection
    connection = module.client('ses', retry_decorator=AWSRetry.jittered_backoff)

    # Usage something like
    connection.list_identity_policies(Identity=identity, aws_retry=True)

    # or 
    connection.list_identity_policies(Identity=identity, aws_retry={retries:5, delay:5})

Or possibly even just passing the whole retry function as a parameter to the call e.g.

    # Setup Connection and opt into the decorator
    connection = module.client('ses', retry_decorator_enabled=True)

    # Usage something like
    connection.list_identity_policies(Identity=identity, retry_decorator=AWSRetry.jittered_backoff())

Pretty sure I could make either of those syntaxes work, and they'd make it explicit while still keeping the benefit of not having to pull out a method for each callsite.

Comment (ryansb) [reply lib/ansible/module_utils/aws/core.py]: Instead of accepting the backoff decorator like `connection = module.client('ses', retry_decorator=AWSRetry.jittered_backoff)` I think you should take the generated decorator, like `connection = module.client('ses', retry_decorator=AWSRetry.jittered_backoff(retries=17))` so that users can change their retry settings per-client, and then on all subsequent calls take a boolean like `aws_retry=True` (defaulted to False) so users have to opt-in at the call site to using retries, but only have to do the configuration once. 

This seems like the good kind of magic to me ;)

Comment (orthanc) [reply lib/ansible/module_utils/aws/core.py]: ~~@ryansb unless I'm miss-understanding you, it's already taking the parameters I just happen to be using the defaults in this particular case. Function call is made [here](https://github.com/ansible/ansible/pull/36623/files#diff-ed4981862e93fb5f5c08b676cfe95aa8R183) and can take parameters.~~ _Edit:_ I get it, you're referring to my proposed syntax. Yes, I agree, keeping the parameters on client creation is better.

I like the idea of adding a boolean "do-retry" at each call-site to keep it explicit, I'll add that in.

Comment (ryansb) [reply lib/ansible/module_utils/aws/core.py]: Yup, you got it. :1st_place_medal:

File: test/integration/targets/aws_ses_identity_policy/tasks/main.yaml [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        identity: "{{ domain_identity }}"
+        state: present
+        <<: *aws_connection_info
+      register: identity_info
+
+    - name: register identity policy
+      aws_ses_identity_policy:
+        identity: "{{ domain_identity }}"
+        policy_name: "{{ policy_name }}"
+        policy: "{{ lookup('template', 'policy.json.j2') }}"
```
Comment (ryansb) [inline test/integration/targets/aws_ses_identity_policy/tasks/main.yaml:242]: In addition to the `lookup` method of taking in the policy, can you also try a policy argument like: 

```
policy:
  Id: SampleAuthorizationPolicy
  Version: "2012-10-17"
  Statement:
  - Sid: DenyAll
    Effect: Deny
    Resource: "{{ identity_info.identity_arn }}"
    Principal: "*"
    Action: "*"
```

The JSON argument type should automatically take that as well, but I'd like to have that validated in the module tests to enforce that UI.

**Comment by orthanc** [comment]
ready_for_review

I believe the new commits cover off all the points raised in review comments specifically:

* The RetryingWrapper is now a new style class
* The Retry behaviour is now opt-in per call-site by specifying an additional keyword arg on the method call `aws_retry=True`
* There's an additional test covering the inline json policy

File: test/integration/targets/aws_ses_identity_policy/tasks/main.yaml (hunk 5 above)
Comment (orthanc) [reply test/integration/targets/aws_ses_identity_policy/tasks/main.yaml:242]: That's cool, I didn't even realize that was possible.

New test added, but I've put it further up the file with the rest of the register variants.

File: lib/ansible/module_utils/aws/core.py (hunk 3 above)
Comment (orthanc) [reply lib/ansible/module_utils/aws/core.py]: Implementation has been updated to @ryansb 's suggestion.

i.e.
* accept the generated decorator when obtaining a connection so we can specify the retry count and other parameters in one place
* at each call-site specify `aws_retry=True` to opt in to the retrying behaviour. If this arg is false or omitted use the unwrapped boto function so there's no retry

File: test/integration/targets/aws_ses_identity_policy/tasks/main.yaml (hunk 5 above)
Comment (ryansb) [reply test/integration/targets/aws_ses_identity_policy/tasks/main.yaml:242]: Thank you!

File: lib/ansible/module_utils/aws/core.py [hunk 6]
```diff
@@ -118,15 +119,17 @@ def debug(self, *args, **kwargs):
...
                           region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        return conn if retry_decorator is None else _RetryingBotoClientWrapper(conn, retry_decorator)
 
-    def resource(self, service):
+    def resource(self, service, retry_decorator=None):
         region, ec2_url, aws_connect_kwargs = get_aws_connection_info(self, boto3=True)
-        return boto3_conn(self, conn_type='resource', resource=service,
+        conn = boto3_conn(self, conn_type='resource', resource=service,
                           region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        return conn if retry_decorator is None else _RetryingBotoClientWrapper(conn, retry_decorator)
```
Comment (ryansb) [inline lib/ansible/module_utils/aws/core.py]: This seems to cause problems with the `Resource` option because of the way resource-clients operate, here's a sample session:

```
>>> import ansible.module_utils.ec2 as aws_ec2      
>>> import ansible.module_utils.aws.core as aws_core
>>> s = boto3.Session(profile_name='ansible-test')       
>>> rr = aws_core._RetryingBotoClientWrapper(s.resource('ec2'), aws_ec2.AWSRetry.jittered_backoff())     
>>> rr.KeyPair('hornet')  
Traceback (most recent call last):                  
  File "<stdin>", line 1, in <module>               
  File "/tmp/ansible/lib/ansible/module_utils/aws/core.py", line 177, in deciding_wrapper                
    return retrying_wrapper(*args, **kwargs)        
  File "/tmp/ansible/lib/ansible/module_utils/cloud.py", line 153, in retry_func                         
    raise e               
ValueError: Required parameter name not set         
Required parameter name not set                     
>>> rr = aws_core._RetryingBotoClientWrapper(s.client('ec2'), aws_ec2.AWSRetry.jittered_backoff())       
>>> rr.describe_key_pairs()                         
{u'KeyPairs': [{....}], 'ResponseMetadata': {'RetryAttempts': 0, 'HTTPStatusCode': 200, 'RequestId': '2fb5545f-3c4e-4cbb-a614-13a4bc2b50ea', 'HTTPHeaders': {'date': 'Thu, 05 Apr 2018 13:21:36 GMT', 'content-length': '954', 'content-type': 'text/xml;charset=UTF-8', 'server': 'AmazonEC2'}}}          
```

the `.client` version of this retry works like a dream, and even help docs are preserved such as `help(rr.describe_key_pairs)`, but I think we should skip rolling this out for the `.resource` version. The naming is fine since the decorator is already `_RetryingBotoClientWrapper` so in the future we could add a `_RetryingBotoResourceWrapper` and the difference would be obvious & easy to spot.

**Review by ryansb — COMMENTED** [review]
One more thing on the retrying system, but the module itself looks great :+1:

File: lib/ansible/module_utils/aws/core.py (hunk 6 above)
Comment (orthanc) [reply lib/ansible/module_utils/aws/core.py]: I'm going to have to dig in to that.

I'll remove the resource version from this PR, once I've done a litle more digging into how the boto resources work I'll see if I can add a resource version as well in a seperate PR.

**Review by ryansb — APPROVED** [review]
